core/src/eval/mod.rs

```diff
@@ -966,7 +969,9 @@ impl<R: ImportResolver, C: Cache> VirtualMachine<R, C> {
                                 attrs.pending_contracts.iter().cloned(),
                                 t.pos,
                             );
-                            inner(slf, acc, value_with_ctr);
+                            if let Some(r) = recursion_limit.checked_sub(1) {
```

Comment:
nitpick: if we test at the very beginning of the function that `recursion_limit > 0` (and return immediatly instead), we have to perform the check only once, instead of at each call site (using saturating sub instead of checked_sub)